server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -109,11 +109,19 @@ public StoredContext stashContext() {
          * This is needed so the DeprecationLogger in another thread can see the value of X-Opaque-ID provided by a user.
          * Otherwise when context is stash, it should be empty.
          */
-        if (context.requestHeaders.containsKey(Task.X_OPAQUE_ID)) {
-            ThreadContextStruct threadContextStruct =
-                DEFAULT_CONTEXT.putHeaders(Map.of(Task.X_OPAQUE_ID, context.requestHeaders.get(Task.X_OPAQUE_ID)));
+
+        if (context.requestHeaders.containsKey(Task.X_OPAQUE_ID) || context.requestHeaders.containsKey(Task.TRACE_PARENT)) {
+            var map = new HashMap<>(Map.<String, String>of());
```

Comment:
Avoids that the default table size of 16 is used when the map can only every hold 2 entries -> reduces allocations.
```suggestion
            Map<String, String> map = new HashMap<>(2, 1);
```